### PR TITLE
Course

### PR DESCRIPTION
**Pull request by James-Koval**
Course functionality was created by me for realizing its in our project.

### CODE REVIEW

File: META-INF/persistence.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (Forfend) [inline META-INF/persistence.xml:1]: Why do we need this?

File: pom.xml [hunk 2]
```diff
@@ -37,11 +37,25 @@
...
+            <groupId>org.modelmapper</groupId>
+            <artifactId>modelmapper</artifactId>
+            <version>2.4.4</version>
+        </dependency>
         <dependency>
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (Forfend) [inline pom.xml:50]: hibernate should already be in your classpath

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import ita.softserve.course_evaluation.entity.Course;
+import ita.softserve.course_evaluation.exception.CourseNotFoundException;
+import ita.softserve.course_evaluation.repository.CourseRepository;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/courses/")
+//@RequiredArgsConstructor
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:13]: remove comment

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package ita.softserve.course_evaluation.controller;
+
+import ita.softserve.course_evaluation.dto.CourseDto;
+import ita.softserve.course_evaluation.entity.Course;
+import ita.softserve.course_evaluation.exception.CourseNotFoundException;
+import ita.softserve.course_evaluation.repository.CourseRepository;
+import org.springframework.web.bind.annotation.*;
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:7]: do not use wildcards

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/courses/")
+//@RequiredArgsConstructor
+public class CourseController {
+    private final CourseRepository courseRepository;
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:15]: Where is CourseService?

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    private Object CourseNotFoundException;
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:16]: What is this?

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    public CourseController(CourseRepository courseRepository) {
+        this.courseRepository = courseRepository;
+    }
+
+    @PostMapping
+    public void postCourse(@RequestBody CourseDto courseDto) {
+        Course course = new Course();
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:24]: It's better to do mapping on service layer

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @GetMapping("/{id}")
+    public Course getCourseById(@PathVariable int id) {
+        return courseRepository.findById(id).orElseThrow(CourseNotFoundException::new);
+    }
+
+    @GetMapping
+    public List<Course> getCourses() {
+        return (List<Course>) courseRepository.findAll();
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:50]: Controller should not communicate with repository layer

File: src/main/java/ita/softserve/course_evaluation/controller/CourseController.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        Course course = new Course();
+        course.setCourseName(courseDto.getCourseName());
+        course.setDescription(courseDto.getDescription());
+        course.setStartDate(courseDto.getStartDate());
+        course.setEndDate(courseDto.getEndDate());
+        courseRepository.save(course);
+    }
+
+    @PutMapping("/{id}")
+    public void putCourse(@PathVariable int id, @RequestBody CourseDto courseDto) {
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/controller/CourseController.java:33]: ```updateCourse``` is better name

File: src/main/java/ita/softserve/course_evaluation/dto/CourseDto.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import ita.softserve.course_evaluation.entity.Course;
+import ita.softserve.course_evaluation.entity.User;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.Objects;
+
+//@Data
+//@AllArgsConstructor
+//@NoArgsConstructor
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/dto/CourseDto.java:12]: Uncomment this or remove

File: src/main/java/ita/softserve/course_evaluation/dto/dtoMapper/CourseDtoMapper.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    public Course toEntity(CourseDto courseDto) {
+        return Objects.isNull(courseDto) ? null : mapper.map(courseDto, Course.class);
+    }
+
+    public static CourseDto toDto(Course course) {
+        return Objects.isNull(course) ? null : mapper.map(course, CourseDto.class);
+    }
+
+    public List<CourseDto> toDto(List<Course> courses) {
+        CourseDtoMapper courseDtoMapper = new CourseDtoMapper();
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/dto/dtoMapper/CourseDtoMapper.java:30]: Probably you don't need this object here

File: src/main/java/ita/softserve/course_evaluation/entity/Course.java [hunk 12]
```diff
@@ -1,36 +1,51 @@
 package ita.softserve.course_evaluation.entity;
 
+import com.sun.istack.NotNull;
+import lombok.NoArgsConstructor;
+
 import javax.persistence.*;
 import java.io.Serializable;
 import java.util.Date;
+
+//@NoArgsConstructor
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/entity/Course.java:10]: Comment

File: src/main/java/ita/softserve/course_evaluation/exception/CourseNotFoundException.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package ita.softserve.course_evaluation.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.NOT_FOUND)
+public class CourseNotFoundException extends RuntimeException {
+    public CourseNotFoundException() {
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/exception/CourseNotFoundException.java:8]: It's good to have a field with description about exception

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 14]
```diff
@@ -1,24 +1,63 @@
...
+import java.util.List;
+import java.util.Optional;
 
 @Service
 public class CourseServiceImpl implements CourseService {
 
-    @Autowired
-    CourseRepository courseRepository;
+    private final CourseRepository courseRepository;
+    private final CourseDtoMapper courseDtoMapper = new CourseDtoMapper();
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:20]: I believe it's available for being injected by Spring Framework

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 15]
```diff
@@ -1,24 +1,63 @@
...
+
+
+    public CourseServiceImpl(CourseRepository courseRepository) {
+        this.courseRepository = courseRepository;
+    }
+
+    @Override
+    public Course addCourse(CourseDto course) {
+        String name = course.getCourseName();
+        courseRepository.findByCourseName(name).ifPresent(n -> {
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:30]: ```courseRepository.findByCourseName(course.getCourseName())...``` - think about memory usage

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 16]
```diff
@@ -1,24 +1,63 @@
...
+            throw new RuntimeException(name + ": this course already created");
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:31]: It's better to have specific exceptions for this case like CourseAlreadyExistsException

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 17]
```diff
@@ -1,24 +1,63 @@
...
+        });
+        return courseRepository.save(courseDtoMapper.toEntity(course));
+    }
+
+    @Override
+    public void delete(int id) {
+        courseRepository.delete(courseDtoMapper.toEntity(getById(id)));
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:38]: It's better to deleteById

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 18]
```diff
@@ -1,24 +1,63 @@
...
+    }
 
+    @Override
+    public Optional<Course> getByName(String name) {
+        return courseRepository.findByCourseName(name);
+    }
+
+    @Override
+    public CourseDto editCourse(CourseDto courseDto) {
+        getById(courseDto.getId());
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:53]: You don't use the result of executing this method

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 19]
```diff
@@ -1,24 +1,63 @@
...
+        return CourseDtoMapper.toDto(courseRepository.save(courseDtoMapper.toEntity(courseDto)));
     }
+
+    @Override
+    public List<CourseDto> getAll() {
+        List<CourseDto> courses = courseDtoMapper.toDto((List<Course>) courseRepository.findAll());
+        return courses.isEmpty() ? new ArrayList<>() : courses;
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:60]: ```Collections.emptyList()```

File: src/main/resources/hibernate.cfg.xml [hunk 20]
```diff
@@ -0,0 +1,15 @@
+<?xml version='1.0' encoding='utf-8'?>
```
Comment (Forfend) [inline src/main/resources/hibernate.cfg.xml:1]: You don't need this file

**Review by Forfend — CHANGES_REQUESTED** [review]

File: META-INF/persistence.xml (hunk 1 above)
Comment (James-Koval) [reply META-INF/persistence.xml:1]: I have no idea because it's an auto-generated script or did you mean about this file in general?

Comment (Forfend) [reply META-INF/persistence.xml:1]: I believe this file is obsolete

File: src/main/java/ita/softserve/course_evaluation/dto/dtoMapper/CourseDtoMapper.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
+package ita.softserve.course_evaluation.dto.dtoMapper;
+
+import ita.softserve.course_evaluation.dto.CourseDto;
+import ita.softserve.course_evaluation.entity.Course;
+import org.springframework.stereotype.Component;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+@Component
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/dto/dtoMapper/CourseDtoMapper.java:10]: You don't need this class as a Bean

File: src/main/java/ita/softserve/course_evaluation/dto/dtoMapper/CourseDtoMapper.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+public class CourseDtoMapper {
+
+    public CourseDtoMapper() {  }
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/dto/dtoMapper/CourseDtoMapper.java:13]: make it private

File: src/main/java/ita/softserve/course_evaluation/entity/Course.java [hunk 23]
```diff
@@ -1,36 +1,49 @@
 package ita.softserve.course_evaluation.entity;
 
+import com.sun.istack.NotNull;
+
 import javax.persistence.*;
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/entity/Course.java:5]: wildcard 🥲🙃

File: src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java [hunk 24]
```diff
@@ -1,24 +1,65 @@
...
+    public CourseServiceImpl(CourseRepository courseRepository) {
+        this.courseRepository = courseRepository;
+    }
 
-    @Transactional
     @Override
-    public void addCourse() {
-        Course courseJava = new Course();
+    public Course addCourse(CourseDto courseDto) {
+        Optional<Course> course = courseRepository.findByCourseName(courseDto.getCourseName());
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation/service/impl/CourseServiceImpl.java:29]: Let's do just ```courseRepository.save(CourseDtoMapper.toEntity(courseDto))```

**Review by Forfend — CHANGES_REQUESTED** [review]

**Review by Forfend — APPROVED** [review]

**Review by Forfend — APPROVED** [review]

**Pull request merged**